WorkspaceServer.Tests/PipelineStepTests.cs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using FluentAssertions.Extensions;
+using WorkspaceServer.Packaging;
+using Xunit;
+
+namespace WorkspaceServer.Tests
+{
+    public class PipelineStepTests
+    {
+        [Fact]
+        public async Task It_produces_a_new_value_when_there_is_none()
+        {
+            var producer = new PipelineStep<string>(() => Task.FromResult("something"));
+            var value = await producer.GetLatestAsync();
+            value.Should().Be("something");
+        }
+
+        [Fact]
+        public async Task It_produces_a_new_value_when_invalidated()
+        {
+            var seed = 0;
+            var producer = new PipelineStep<int>(() => Task.FromResult(Interlocked.Increment(ref seed)));
+            var value = await producer.GetLatestAsync();
+            value.Should().Be(1);
+
+            producer.Invalidate();
+            var newValue = await producer.GetLatestAsync();
+            newValue.Should().Be(2);
+        }
+
+        [Fact]
+        public async Task It_propagates_exception()
+        {
+            var producer = new PipelineStep<int>( () => throw new InvalidOperationException());
+            producer.Awaiting(p => p.GetLatestAsync())
+                .Should()
+                .Throw<InvalidOperationException>();
+           
+        }
+
+        [Fact]
+        public async Task It_remains_invalid_if_exceptions_are_thrown()
+        {
+            var seed = 0;
+            var producer = new PipelineStep<int>(() =>
+            {
+                var next = Interlocked.Increment(ref seed);
+                if (next == 1)
+                {
+                    throw new InvalidOperationException();
+                }
+                return Task.FromResult(next);
+            });
+
+            producer.Awaiting(p => p.GetLatestAsync())
+                .Should()
+                .Throw<InvalidOperationException>();
+
+            var value = await producer.GetLatestAsync();
```

Comment:
Maybe I'm misunderstanding this, but the test is called "remains invalid", so shouldn't this invocation also throw?

Reply:
I think the meaning of "invalid" is unclear here. What the test is testing is that the pipeline remains in an invalidated state, so that the next call to `GetLatestAsync` retries the work that failed due to the exception.

---
In reply to: [287448147](https://github.com/dotnet/try/pull/236#discussion_r287448147) [](ancestors = 287448147)